### PR TITLE
Support `__bool__` where sensible

### LINKED ISSUE #2691: Support `__bool__` in some cases?
This came up from a user question. Currently we cannot use `if sc.scalar(1, unit='m') < sc.scalar(2, unit='m')`.

#2529 implemented `__float__` and `__int__`. The conditions there are:

- Only supporting `Variable` (not `DataArray`).
- No variance.
- 0-D, `shape=(1, )` is *not* ok.
- `unit=1` or `unit=None`.

Are the same conditions sensible for `__bool__`? I am unsure about support for `unit=1`, but deviating may also add confusion?

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #2691.

### CODE REVIEW

File: lib/python/bind_operators.h [hunk 1]
```diff
@@ -28,7 +28,13 @@ template <class T, class... Ignored>
 void bind_common_operators(pybind11::class_<T, Ignored...> &c) {
   c.def("__abs__", [](const T &self) { return abs(self); });
   c.def("__repr__", [](const T &self) { return to_string(self); });
-  c.def("__bool__", [](const T &) {
+  c.def("__bool__", [](const T &self) {
+    if constexpr (std::is_same_v<T, scipp::Variable>) {
+      if (self.unit() != scipp::units::none)
+        throw scipp::except::UnitError(
+            "The thruth value of a variable with unit is undefined.");
```
Comment (nvaytet) [inline lib/python/bind_operators.h:35]:
```suggestion
            "The truth value of a variable with unit is undefined.");
```

File: lib/python/bind_operators.h [hunk 2]
```diff
@@ -28,7 +28,13 @@ template <class T, class... Ignored>
...
+      return self.template value<bool>() == true;
```
Comment (nvaytet) [inline lib/python/bind_operators.h:36]: Is this the part that makes sure we don't have more than one value in the `self`? `self.value` raises if the variable is not 0-D?

File: tests/variable_scalar_test.py [hunk 3]
```diff
@@ -89,7 +89,45 @@ def test_scalar_Variable_conversion_to_builtin_float_with_variance():
...
+def test_scalar_Variable_conversion_to_builtin_bool_True():
+    assert sc.scalar(True)
+
+
+def test_scalar_Variable_conversion_to_builtin_bool_False():
+    assert not sc.scalar(False)
+
+
+def test_scalar_Variable_conversion_to_builtin_bool_bad_dtype():
+    var = sc.scalar(value=1.0, unit=None)
```
Comment (nvaytet) [inline tests/variable_scalar_test.py:107]: Out of interest, do we support conversion from integer? e.g. `sc.scalar(value=1, unit=None)`

**Review by jl-wynen — CHANGES_REQUESTED** [review]

File: lib/python/bind_operators.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply lib/python/bind_operators.h:36]: `bool` cannot have variances.

[the remark this answers is not in the document]
Comment (jl-wynen) [reply lib/python/bind_operators.h:36]: Right

**Review by jl-wynen — APPROVED** [review]

**Pull request merged**